python/pyabacus/src/py_diago_dav_subspace.cpp

```diff
@@ -0,0 +1,88 @@
+#include <complex>
+#include <functional>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <pybind11/complex.h>
+#include <pybind11/numpy.h>
+
+#include "module_hsolver/diago_dav_subspace.h"
+#include "module_hsolver/kernels/math_kernel_op.h"
+#include "module_base/module_device/types.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+void bind_diago_dav_subspace(py::module& m)
+{
+    py::module hsolver = m.def_submodule("hsolver");
+
+    py::class_<hsolver::diag_comm_info>(hsolver, "diag_comm_info")
+        .def(py::init<const int, const int>(), "rank"_a, "nproc"_a)
+        .def_readonly("rank", &hsolver::diag_comm_info::rank)
+        .def_readonly("nproc", &hsolver::diag_comm_info::nproc);
+
+    py::class_<hsolver::Diago_DavSubspace<std::complex<double>, base_device::DEVICE_CPU>>(hsolver, "Diago_DavSubspace")
+        .def(py::init<const std::vector<double>, const int, const int,
+                    const int, const double, const int, const bool,
+                    const hsolver::diag_comm_info>(),
+        "precondition"_a, "nband"_a, "nbasis"_a, "david_ndim"_a, "diag_thr"_a, "diag_nmax"_a, "need_subspace"_a, "diag_comm"_a)
+        .def("diag",
+            [] (hsolver::Diago_DavSubspace<std::complex<double>, base_device::DEVICE_CPU>& self,
+                py::array_t<std::complex<double>> h_mat,
+                py::array_t<std::complex<double>> psi_in,
+                const int psi_in_dmax,
+                py::array_t<double> eigenvalue_in,
+                const std::vector<bool>& is_occupied,
+                const bool& scf_type)
```

Comment:
Good. However, Davidson Method should be designed as a pure math class(function). It means you can call the ABACUS API in your inner codes but expose the python function as a pure math class. Plz see the [Davidson implementation](https://github.com/pyscf/pyscf/blob/master/pyscf/lib/linalg_helper.py#L200-L225) in PySCF.

Reply:
Also, plz add `rdoc` to show the docs in pyabacus.

Reply:
i have added `pbdoc` in pyabacus and designed new interface